domain/storage/service/storagepool.go

```diff
@@ -0,0 +1,246 @@
+	// Below is the code from state that will need to be ported.
+	/*
+		var inUse bool
+		cfg, err := sb.config(context.Background())
+		if err != nil {
+			return errors.Trace(err)
+		}
+		operatorStorage, ok := cfg.AllAttrs()[k8sconstants.OperatorStorageKey]
+		if sb.modelType == ModelTypeCAAS && ok && operatorStorage == poolName {
+			apps, err := sb.allApplications()
+			if err != nil {
+				return errors.Trace(err)
+			}
+			inUse = len(apps) > 0
+		} else {
+			query := bson.D{{"constraints.pool", bson.D{{"$eq", poolName}}}}
+			pools, err := storageCollection.Find(query).Count()
+			if err != nil {
+				return errors.Trace(err)
+			}
+			inUse = pools > 0
+		}
+		if inUse {
+			return errors.Errorf("storage pool %q in use", poolName)
+		}
+	*/
+	err := s.st.DeleteStoragePool(ctx, name)
+	return errors.Annotatef(err, "deleting storage pool %q", name)
+}
+
+// ReplaceStoragePool replaces an existing storage pool, returning an error
+// satisfying [storageerrors.PoolNotFoundError] if a pool with the name does not exist.
+func (s *StoragePoolService) ReplaceStoragePool(ctx context.Context, name string, providerType storage.ProviderType, attrs PoolAttrs) error {
+	// Use the existing provider type unless explicitly overwritten.
+	if providerType == "" {
+		existingConfig, err := s.st.GetStoragePoolByName(ctx, name)
+		if err != nil {
+			return errors.Trace(err)
+		}
+		providerType = storage.ProviderType(existingConfig.Provider)
+	}
+
+	err := s.validateConfig(name, providerType, attrs)
+	if err != nil {
+		return errors.Trace(err)
+	}
+
+	attrsToSave := transform.Map(attrs, func(k string, v any) (string, string) { return k, fmt.Sprint(v) })
+	sp := domainstorage.StoragePoolDetails{
+		Name:     name,
+		Provider: string(providerType),
+		Attrs:    attrsToSave,
+	}
+	err = s.st.ReplaceStoragePool(ctx, sp)
+	return errors.Annotatef(err, "replacing storage pool %q", name)
+}
+
+// ListStoragePools returns the storage pools matching the specified filter.
+func (s *StoragePoolService) ListStoragePools(ctx context.Context, filter domainstorage.StoragePoolFilter) ([]*storage.Config, error) {
+	if err := s.validatePoolListFilter(filter); err != nil {
+		return nil, errors.Trace(err)
+	}
+
+	pools, err := domainstorage.BuiltInStoragePools()
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+
+	sp, err := s.st.ListStoragePools(ctx, filter)
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+	pools = append(pools, sp...)
+
+	results := make([]*storage.Config, len(pools))
+	for i, p := range pools {
+		results[i], err = s.toStorageConfig(p)
+		if err != nil {
+			return nil, errors.Trace(err)
+		}
+	}
+	return results, nil
+}
+
+func (s *StoragePoolService) validatePoolListFilter(filter domainstorage.StoragePoolFilter) error {
+	if err := s.validateProviderCriteria(filter.Providers); err != nil {
+		return errors.Trace(err)
+	}
+	if err := s.validateNameCriteria(filter.Names); err != nil {
+		return errors.Trace(err)
+	}
```

Comment:
These could be replaceed with `ListStoragePoolsByProviders` and `ListStoragePoolsByName`.

Reply:
Replaced filter with individual args for query terms